solver/src/analytics.rs

```diff
@@ -0,0 +1,123 @@
+use crate::{metrics::SolverMetrics, settlement::Settlement};
+use bigdecimal::{ToPrimitive, Zero};
+use ethcontract::H160;
+use model::order::OrderUid;
+use num::BigRational;
+use shared::conversions::U256Ext;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    fmt::{Display, Formatter},
+    sync::Arc,
+};
+
+#[derive(Clone)]
+struct SurplusInfo {
+    solver_name: &'static str,
+    absolute: BigRational,
+    ratio: BigRational,
+}
+
+impl Display for SurplusInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Surplus {{solver: {}, absolute: {:.2e}, ratio: {:.2e} }}",
+            self.solver_name,
+            self.absolute.to_f64().unwrap_or(f64::NAN),
+            self.ratio.to_f64().unwrap_or(f64::NAN)
+        )
+    }
+}
+
+fn get_prices(settlement: &Settlement) -> HashMap<&H160, BigRational> {
+    settlement
+        .clearing_prices()
+        .iter()
+        .map(|(token, price)| (token, price.to_big_rational()))
```

Comment:
ubernit: Since `H160` is a `Copy` type, I think it makes slightly more sense to:
```suggestion
fn get_prices(settlement: &Settlement) -> HashMap<H160, BigRational> {
    settlement
        .clearing_prices()
        .iter()
        .map(|(token, price)| (*token, price.to_big_rational()))
```

This may also simplify lifetime things down the line, and it makes things a little less awkward when getting items (you don't need `&&H160`s). For example no need to take the reference of the sell token address on line 53:
```rust
            let sell_token_price = &submitted_prices[trade.order.order_creation.sell_token];
```

Reply:
I was able to make your suggested change alright, but was not able to remove the reference when accessing the price map.